src/credentialstore/credentialstore.ts

```diff
@@ -26,44 +44,141 @@ export class CredentialStore implements ICredentialStore {
-            self._client
-            .sendRequest(Contracts.SaveCredentialRequest.type, cred)
-            .then(status => {
-                resolve(status);
-            }, err => reject(err));
-        });
+        try {
+            let success: boolean;
+            // Use native credential if the setting is on
+            if (this._useNativeCredentials) {
+                success = await this.nativeSaveCredential(credentialId, cred);
+            } else {
+                success = await this._client.sendRequest(Contracts.SaveCredentialRequest.type, cred);
+            }
+            return success;
+        } catch (err) {
+            throw(err);
+        }
     }
 
-    public deleteCredential(credentialId: string): Promise<boolean> {
-        let self = this;
+    /**
+     * Removes a credential from the credential store
+     * @param credentialId the ID uniquely identifying this credential
+     */
+    public async deleteCredential(credentialId: string): Promise<boolean> {
         let cred: Contracts.Credential = new Contracts.Credential();
         cred.credentialId = credentialId;
-        return new Promise<boolean>( (resolve, reject) => {
-            self._client
-            .sendRequest(Contracts.DeleteCredentialRequest.type, cred)
-            .then(status => {
-                resolve(status);
-            }, err => reject(err));
-        });
+        try {
+            let success: boolean;
+            // Use native credential if the setting is on
+            if (this._useNativeCredentials) {
+                success = await this.nativeDeleteCredential(credentialId);
+            } else {
+                success = await this._client.sendRequest(Contracts.DeleteCredentialRequest.type, cred);
+            }
+            return success;
+        } catch (err) {
+            throw(err);
+        }
+    }
+
+    // Native Credential implementations
+
+    private async nativeReadCredential(credentialId: string, cred: Contracts.Credential): Promise<Contracts.Credential> {
+        try {
+            const savedPassword: string = await this._context.secrets.get(credentialId);
+            cred.password = savedPassword;
+            return cred;
+        } catch (err) {
+            throw(err);
+        }
+    }
+
+    private async nativeSaveCredential(credentialId: string, password: any): Promise<boolean> {
+        try {
+            await this._context.secrets.store(credentialId, password);
+            const savedPassword = await this._context.secrets.get(credentialId);
+            if (savedPassword === password) {
+                return true;
+            }
+            return false;
+        } catch (err) {
+            throw(err);
+        }
+    }
+
+    private async nativeDeleteCredential(credentialId: string): Promise<boolean> {
+        try {
+            await this._context.secrets.delete(credentialId);
+            const savedPassword = await this._context.secrets.get(credentialId);
```

Comment:
Why do you need to check this? We should generally trust their APIs to do what they say at runtime.